tensorflow/python/ops/image_ops_impl.py

```diff
@@ -3555,3 +3558,55 @@ def combined_non_max_suppression(boxes,
     return gen_image_ops.combined_non_max_suppression(
         boxes, scores, max_output_size_per_class, max_total_size, iou_threshold,
         score_threshold, pad_per_class)
+@tf_export('image.median_filter_2D')
+def median_filter_2D(input,filter_shape=(3,3)):
+    """  This methods takes 3D Tensor Images.
+         Other than Tensor it takes optional parameter filter_Size
+         Default Filter Shape = (3 , 3)
+         This Median Filtering is done by using 2D filters of user's choice
+         Filter_size should be odd
+         This method takes both kind of images where pixel values lie between 0 to 255 and where it lies between 0.0 and 1.0
+    """
+
+    input = image_ops_impl._Assert3DImage(input)
+    m,no,ch = int(input.shape[0]),int(input.shape[1]),int(input.shape[2])
+    filter_shapex = filter_shape[0]
+    filter_shapey = filter_shape[1]
+    if m < filter_shapex or no < filter_shapey:
+        raise ValueError("No of Pixels in each dimension of the image should be more than the filter size. Got filter_shape "
+                         "(%sx" % filter_shape[0]+"%s)."%filter_shape[1] +" Image Shape (%s)"% input.shape)
+    if filter_shapex % 2 == 0 or filter_shapey % 2 == 0:
+        raise ValueError("Filter size should be odd. Got filter_shape (%sx" % filter_shape[0]+"%s)"%filter_shape[1] )
+    input = math_ops.cast(input,dtypes.float64)
+    def my_func (input2):
+        tf_i = input2.reshape(m*no*ch)
+        maxi = max(tf_i)
+        if maxi <= 1:
+            input2 /= 1
+        else :
+            input2 /= 255
+        #k and l is the Zero-padding size
+        res = np.empty((m,no,ch))
+        for a in range(ch):
+            img = input2[:,:,a:a+1]
+            img = img.reshape(m,no)
+            k = filter_shapex - 1
+            l = filter_shapey - 1
+            img  = np.pad(img,((k / 2, k / 2), (l / 2,l / 2)),'constant', constant_values=(0, 0))
+            res1 = np.empty((m,no))
+            for i in range(img.shape[0] - k) :
+                for j in range(img.shape[1] - l) :
+                    li = []
+                    for b in range(i, i + filter_shapex):
+                        for d in range(j, j + filter_shapey):
+                            li.append(img[b][d])
+                    li.sort()
+                    res1[i][j] = li[len(li) / 2]
+            res1 = res1.reshape(m,no,1)
+            res[:,:,a:a+1] = res1
+        res *= 255
+        res = res.astype('int64')
+        return res
+
+    y = script_ops.py_func(my_func, [input], dtypes.int64)
```

Comment:
Why using a py_func instead of computing means via tf.nn.top_k?

Reply:
I am contributing to tensorflow for the first time, Thanks for your detailed help. 
As per your suggestion in pull request #26741 . You gave me two options. One of them is using py_func. I have checked for both, and found py_func more convenient to implement the logic without using session . Please let me know if you want more changes. Thanks.

Reply:
We generally do not add operations to the tf namespace which rely on py_func because py_func does not work on lite or distributed setups, and can have performance problems due to moving back to python all the time.

Reply:
```suggestion
   @tf_export('image.median_filter_2D')
def median_filter_2D(input,filter_shape=(3,3)):
    """This method performs Median Filtering on image.Filter shape can be user given.
       This method takes both kind of images where pixel values lie between 0 to 255 and where it lies between 0.0 and 1.0
       Args:
           input: A 3D `Tensor` of type `float32` or 'int32' or 'float64' or 'int64 and of shape`[rows, columns, channels]`

           filter_shape: Optional Argument. A tuple of 2 integers .(R,C).R is the first value is the number of rows in the filter
           and C is the second value in the filter is the number of columns in the filter. This creates a filter of shape (R,C) or
           RxC filter. Default value = (3,3)

        Returns:
            A 3D median filtered image tensor of shape [rows,columns,channels] and type 'int32'. Pixel value of returned tensor
            ranges between 0 to 255
    """
    if not isinstance(filter_shape,tuple):
        raise TypeError("Filter shape must be a tuple")
    if len(filter_shape) != 2:
        raise ValueError("Filter shape must be a tuple of 2 integers .Got %s values in tuple"% len(filter_shape))
    filter_shapex = filter_shape[0]
    filter_shapey = filter_shape[1]
    if isinstance(filter_shapex,int) and isinstance(filter_shapey,int):
        pass
    else :
        raise TypeError("Size of the filter must be Integers")
    input = image_ops_impl._Assert3DImage(input)
    m,no,ch = input.shape[0],input.shape[1],input.shape[2]
    if not m.__eq__(tensor_shape.Dimension(None)) and not no.__eq__(tensor_shape.Dimension(None)) \
            and not ch.__eq__(tensor_shape.Dimension(None)):
        m,no,ch = int(m),int(no),int(ch)
    else :
        raise TypeError("All the Dimensions of the input image tensor must be Integers")
    if m < filter_shapex or no < filter_shapey:
        raise ValueError("No of Pixels in each dimension of the image should be more than the filter size. Got filter_shape "
                         "(%sx" % filter_shape[0]+"%s)."%filter_shape[1] +" Image Shape (%s)"% input.shape)
    if filter_shapex % 2 == 0 or filter_shapey % 2 == 0:
        raise ValueError("Filter size should be odd. Got filter_shape (%sx" % filter_shape[0]+"%s)"%filter_shape[1] )
    input = math_ops.cast(input,dtypes.float32)
    tf_i = array_ops.reshape(input,[m*no*ch])
    ma = math_ops.reduce_max(tf_i)

    def normalize(li):
        one = ops.convert_to_tensor(1.0)
        two = ops.convert_to_tensor(255.0)

        def func1():
            return li

        def func2():
            return math_ops.truediv(li, two)

        return control_flow_ops.cond(gen_math_ops.greater(ma, one), func2, func1)

    input = normalize(input)
    #k and l is the Zero-padding size
    listi = []
    for a in range(ch):
        img = input[:,:,a:a+1]
        img = array_ops.reshape(img,[m,no])
        k = filter_shapex - 1
        l = filter_shapey - 1
        img  = array_ops.pad(img,constant_op.constant([[k / 2, k / 2], [l / 2,l / 2]]),'CONSTANT')
        li = []
        for i in range(img.shape[0] - k) :
            for j in range(img.shape[1] - l) :
                s1 = array_ops.slice(img, [i, j], [filter_shapex, filter_shapey])
                s1 = array_ops.reshape(s1, [-1])
                mid = s1.get_shape()[0] // 2 + 1
                s1 = nn_ops.top_k(s1, mid).values[-1]
                li.append(s1)
        li = array_ops.stack(li)
        li = array_ops.reshape(li, [m, no, 1])
        listi.append(li)
    y = array_ops.concat(listi[0], 2)

    for i in range(len(listi) - 1):
        y = array_ops.concat([y, listi[i + 1]], 2)

    y *= 255
    y = math_ops.cast(y,dtypes.int32)

    return y
```
@alextp .I have written the entire program in tensorflow without using py_func or eager_py_func in this suggestion. I have witnessed this above suggestion is consuming more memory and time . There is huge performance loss. This solution is near about taking 100 * time with respect to py_func solution. Still, as you have asked me to come up with an alternative all tensorflow solution, i have uploaded the solution as suggestion. I did not commit this code. If you think, the code should be committed , please ask me. Apparently, high time usage is due to median calculation by top_k.

Reply:
Thanks for the rewrite!

The way this code is written now it's doing many many calls to top_k. Instead I think you can stack all the arguments to all calls to top_k and do a single call to top_k which returns the top_k per row of the stacked matrix. This will probably be much more efficient.

Can you give it a try?

Reply:
```
@tf_export('image.median_filter_2D')
def median_filter_2D(input,filter_shape=(3,3)):
    """This method performs Median Filtering on image.Filter shape can be user given.
       This method takes both kind of images where pixel values lie between 0 to 255 and where it lies between 0.0 and 1.0
       Args:
           input: A 3D `Tensor` of type `float32` or 'int32' or 'float64' or 'int64 and of shape`[rows, columns, channels]`

           filter_shape: Optional Argument. A tuple of 2 integers .(R,C).R is the first value is the number of rows in the filter
           and C is the second value in the filter is the number of columns in the filter. This creates a filter of shape (R,C) or
           RxC filter. Default value = (3,3)

        Returns:
            A 3D median filtered image tensor of shape [rows,columns,channels] and type 'int32'. Pixel value of returned tensor
            ranges between 0 to 255
    """
    if not isinstance(filter_shape,tuple):
        raise TypeError("Filter shape must be a tuple")
    if len(filter_shape) != 2:
        raise ValueError("Filter shape must be a tuple of 2 integers .Got %s values in tuple"% len(filter_shape))
    filter_shapex = filter_shape[0]
    filter_shapey = filter_shape[1]
    if isinstance(filter_shapex,int) and isinstance(filter_shapey,int):
        pass
    else :
        raise TypeError("Size of the filter must be Integers")
    input = image_ops_impl._Assert3DImage(input)
    m,no,ch = input.shape[0],input.shape[1],input.shape[2]
    if not m.__eq__(tensor_shape.Dimension(None)) and not no.__eq__(tensor_shape.Dimension(None)) \
            and not ch.__eq__(tensor_shape.Dimension(None)):
        m,no,ch = int(m),int(no),int(ch)
    else :
        raise TypeError("All the Dimensions of the input image tensor must be Integers")
    if m < filter_shapex or no < filter_shapey:
        raise ValueError("No of Pixels in each dimension of the image should be more than the filter size. Got filter_shape "
                         "(%sx" % filter_shape[0]+"%s)."%filter_shape[1] +" Image Shape (%s)"% input.shape)
    if filter_shapex % 2 == 0 or filter_shapey % 2 == 0:
        raise ValueError("Filter size should be odd. Got filter_shape (%sx" % filter_shape[0]+"%s)"%filter_shape[1] )
    input = math_ops.cast(input,dtypes.float32)
    tf_i = array_ops.reshape(input,[m*no*ch])
    ma = math_ops.reduce_max(tf_i)

    def normalize(li):
        one = ops.convert_to_tensor(1.0)
        two = ops.convert_to_tensor(255.0)

        def func1():
            return li

        def func2():
            return math_ops.truediv(li, two)

        return control_flow_ops.cond(gen_math_ops.greater(ma, one), func2, func1)

    input = normalize(input)
    #k and l is the Zero-padding size
    listi = []
    for a in range(ch):
        img = input[:,:,a:a+1]
        img = array_ops.reshape(img,[1,m,no,1])
        slic = gen_array_ops.extract_image_patches(img,[1, filter_shapex, filter_shapey, 1],
                                                   [1, 1, 1, 1],[1, 1, 1, 1],padding='SAME')
        mid = (filter_shapex * filter_shapey) / 2 + 1
        li = nn_ops.top_k(slic, mid, sorted=True)
        li = sample_stats.percentile(li[0], 50, axis=3)
        li = array_ops.reshape(li, [m, no, 1])
        listi.append(li)
    y = array_ops.concat(listi[0], 2)

    for i in range(len(listi) - 1):
        y = array_ops.concat([y, listi[i + 1]], 2)

    y *= 255
    y = math_ops.cast(y, dtypes.int32)

    return y
```
@alextp Thanks for your detailed suggestions. I think you will now be happy with the current version of the code. Single call to top_k solved the issue of performance. I have also committed the new code. Please review and test this.